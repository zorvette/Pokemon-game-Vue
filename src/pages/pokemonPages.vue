<template>

<h1 v-if="!pokemon" >Espere por favor...</h1>

  <div v-else>
  <h1>¿Quién es este pokémon?</h1>

  <pokemonPicture 
    :pokemonId="pokemon.id" 
    :showPokemon="showPokemon"
  />
  <pokemonOptions 
    :pokemons="pokemonArr"
    @selection="checkAnswer"
  /> 

    <div v-if="showAnswer">
      <h2 class="fade-in"> {{message}} </h2>
      <button @click="newGame" >
        Nuevo Juego
      </button>
    </div>
  </div>

</template>

<script>
import pokemonPicture from '@/components/pokemonPicture.vue'
import pokemonOptions from '@/components/pokemonOptions.vue'

import getPokemonOptions from '@/helpers/getPokemonOptions'

// console.log( getPokemonOptions() )

export default {
  components: { pokemonPicture, pokemonOptions },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: ""
    }
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions()

      const rndInt = Math.floor( Math.random() * 4 )

    this.pokemon = this.pokemonArr[ rndInt ]
    },
    checkAnswer( selectedId ) {
      this.showPokemon = true
      this.showAnswer = true

      if( selectedId === this.pokemon.id ) {
        this.message = `Correcto, ${this.pokemon.name}`
      } else {
        this.message = `Oops, era ${this.pokemon.name}`
      }
    },
    newGame() {
      this.showPokemon = false
      this.showAnswer  = false
      this.pokemon     = null
      this.pokemonArr  = []
      this.mixPokemonArray()
    }
  },
  mounted() {
    this.mixPokemonArray()
  }
}
</script>
