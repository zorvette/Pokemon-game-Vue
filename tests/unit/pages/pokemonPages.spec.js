import { shallowMount, mount } from '@vue/test-utils'
import pokemonPages from '@/pages/pokemonPages'
import { mockPokemon } from '../mocks/pokemons.mock'

let wrapper

beforeEach(() => {
  wrapper = shallowMount( pokemonPages )
})

describe(' pokemonPages Component ', () => {

  test('debe hacer match con el snapshot', () => {
    expect( wrapper.html() ).toMatchSnapshot()
  })

  test('debe de llamar mixPokemonArray al montar', () => {

    const mixPokemonArraySpy = jest.spyOn( pokemonPages.methods, 'mixPokemonArray' )
    const wrapper = shallowMount( pokemonPages )
    expect( mixPokemonArraySpy ).toHaveBeenCalled()

  })

  test('debe hacer match con el snapshot cuando carguen los pokemons', () => {
    const wrapper = shallowMount( pokemonPages, {
      data() {
        return {
          pokemonArr: mockPokemon,
          pokemon: mockPokemon[0],
          showPokemon: false,
          showAnswer: false,
          message: ""
        }
      }
    } )
    expect( wrapper.html() ).toMatchSnapshot()
  })

  test('debe de mostrar los componentes de pokemonPicture y pokemonOptions', () => {

    const wrapper = shallowMount( pokemonPages, {
      data() {
        return {
          pokemonArr: mockPokemon,
          pokemon: mockPokemon[0],
          showPokemon: false,
          showAnswer: false,
          message: ""
        }
      }
    } )

    const picture = wrapper.find('pokemon-picture-stub')
    const options = wrapper.find('pokemon-options-stub')

    expect(picture.exists()).toBe(true)
    expect(options.exists()).toBe(true)

    expect(picture.attributes('pokemonid')).toBe('1')
    expect(options.attributes('pokemons')).toBeTruthy()
  })

  test('pruebas con checkAnswer', async() => {

    const wrapper = shallowMount( pokemonPages, {
      data() {
        return {
          pokemonArr: mockPokemon,
          pokemon: mockPokemon[0],
          showPokemon: false,
          showAnswer: false,
          message: ""
        }
      }
    } )
    await wrapper.vm.checkAnswer(1)
    
    expect(wrapper.find('h2').exists()).toBeTruthy()
    expect( wrapper.vm.showPokemon ).toBeTruthy()
    expect(wrapper.find('h2').text()).toBe(`Correcto, ${mockPokemon[0].name}`)
    
    await wrapper.vm.checkAnswer(10)
    
    expect(wrapper.vm.message).toBe(`Oops, era ${mockPokemon[0].name}`)

  })

})