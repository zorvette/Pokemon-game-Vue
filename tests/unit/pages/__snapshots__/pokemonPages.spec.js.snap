// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` pokemonPages Component  debe hacer match con el snapshot 1`] = `<h1>Espere por favor...</h1>`;

exports[` pokemonPages Component  debe hacer match con el snapshot cuando carguen los pokemons 1`] = `
<div>
  <h1>¿Quién es este pokémon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
